design/authenticators/authn_sut/authn_sut_solution_design.md

```diff
@@ -0,0 +1,384 @@
+| 4  | User or host is not defined in Conjur                                           | '{0-role-name}' wasn't found                                                           |
+| 5  | Login error                               | Login error: ...                                     |
+| 6  | Missing SUT in /authenticate request body | Single use token not provided    |
+| 7  | SUT not found or expired | Single use token is invalid    |
+| 8  | Missing code challenge in /login request headers | Code challenge not provided    |
+| 9  | Invalid code challenge algorithm in /login request headers | Code challenge algorithm not supported    |
+| 10  | Invalid code challenge in /login request headers | Code challenge not valid    |
+| 11  | Missing code verifier in /authenticate request body | Code verifier not provided    |
+| 12  | Code verifier does not match code challenge | Code verifier does not match code challenge    |
+
+###### Debug Log level
+
+|    | Scenario                                              | Log message                                                                            |
+|--- |-------------------------------------------------------|----------------------------------------------------------------------------------------|
+| 1  | After validating SUT | Single use token for {0-role-name} has been validated successfully                |
+
+##### Audit Log Messages
+
+Same as other authenticators - one message for successful authentication, one for failed authentication.
+
+#### Policy
+
+As with other authenticators, the SUT authenticator will be disabled by default.
+It will be enabled by creating an authenticator policy resource in Conjur and
+adding it to the list of enabled authenticators in the Conjur configuration, for
+example in the `CONJUR_AUTHENTICATORS` environment variable.
+
+The authenticator webservice must be declared in Conjur policy:
+
+```yml
+- !policy
+  id: conjur/authn-sut
+  body:
+  - !webservice
+
+  - !group 
+    id: authenticatable
+    annotations:
+      description: Group with permission to authenticate using this authenticator
+
+  - !permit
+    role: !group authenticatable
+    privilege: [ read, authenticate ]
+    resource: !webservice
+```
+
+and the `authenticatable` group can be used to entitle a user(s) to use the
+authenticator:
+
+```yml
+- !grant
+  role: !group conjur/authn-sut/authenticatable
+  member: !user <username>
+```
+
+#### API
+
+We will need two new API endpoints for the SUT authenticator:
+
+##### `GET /authn-sut/<account>/login`
+
+This endpoint will be used to generate a SUT for a user. The user will provide
+their username and password (or API key) in the `Authentication` header using the
+standard format for HTTP Basic authentication.
+
+Additionally, the client will provide a Code Challenge and Code Challenge Algorithm
```

Comment:
Headers are intended to provide metadata about the corresponding body. To me, the hashing algorithm makes sense, but as the code challenge value is critical to the overall authentication request, it should be part of the login body (with the other key bits: username/password or host/API key).

I suspect we're proposing header values to avoid the use of the `POST` action.  In this case, I'd argue that it's more appropriate to use `POST` because the result is the creation of a single-use token.  In the normal login flow, the result is that user's API key, which already exists.

Reply:
Yes, you got it exactly. Use of a header is intended to allow us to use a GET request to be consistent with the authn and authn-ldap routes. I'm fine with making it a POST request, just knowing that it may cause some confusion.

Reply:
What do you think about changing this to a `POST`: 

`/authn-sut/<account>/login`

**Headers**
|Header|Sample Value|
|-|-|
|Code-Challenge-Algorithm|sha256|
|Authentication|Zm9vOmJhcg== (base64 encoding of "foo:bar")|

**Body**
Code-Challenge-Value